File: app/src/main/java/com/hover/stax/onboarding/StaxAutoScrollViewPager.java
package com.hover.stax.onboarding;

import android.content.Context;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.animation.Interpolator;
import android.widget.Scroller;

import androidx.viewpager.widget.PagerAdapter;
import androidx.viewpager.widget.ViewPager;

import org.jetbrains.annotations.NotNull;

import java.lang.ref.WeakReference;
import java.lang.reflect.Field;

public class StaxAutoScrollViewPager extends ViewPager {

	public static final int DEFAULT_INTERVAL = 1500;

	public static final int LEFT = 0;
	public static final int RIGHT = 1;
	public static final int SLIDE_BORDER_MODE_NONE = 0;
	public static final int SLIDE_BORDER_MODE_CYCLE = 1;
	public static final int SLIDE_BORDER_MODE_TO_PARENT = 2;
	private long interval = DEFAULT_INTERVAL;
	private int direction = RIGHT;
	private boolean isCycle = true;
	private boolean stopScrollWhenTouch = true;
	private int slideBorderMode = SLIDE_BORDER_MODE_NONE;
	private boolean isBorderAnimation = true;
	private double autoScrollFactor = 1.0;
	private double swipeScrollFactor = 1.0;

	private Handler handler;
	private boolean isAutoScroll = false;
	private boolean isStopByTouch = false;
	private float touchX = 0f, downX = 0f;
	private CustomDurationScroller scroller = null;

	public static final int SCROLL_WHAT = 0;

	public StaxAutoScrollViewPager(Context paramContext) {
		super(paramContext);
		init();
	}

	public StaxAutoScrollViewPager(Context paramContext, AttributeSet paramAttributeSet) {
		super(paramContext, paramAttributeSet);
		init();
	}

	private void init() {
		handler = new MyHandler(this);
		setViewPagerScroller();
	}


	public void startAutoScroll() {
		isAutoScroll = true;
		sendScrollMessage((long) (interval + scroller.getDuration() / autoScrollFactor * swipeScrollFactor));
	}

	public void startAutoScroll(int delayTimeInMills) {
		isAutoScroll = true;
		sendScrollMessage(delayTimeInMills);
	}


	public void stopAutoScroll() {
		isAutoScroll = false;
		handler.removeMessages(SCROLL_WHAT);
	}

	public void setSwipeScrollDurationFactor(double scrollFactor) {
		swipeScrollFactor = scrollFactor;
	}


	public void setAutoScrollDurationFactor(double scrollFactor) {
		autoScrollFactor = scrollFactor;
	}

	private void sendScrollMessage(long delayTimeInMills) {
		handler.removeMessages(SCROLL_WHAT);
		handler.sendEmptyMessageDelayed(SCROLL_WHAT, delayTimeInMills);
	}


	private void setViewPagerScroller() {
		try {
			Field scrollerField = ViewPager.class.getDeclaredField("mScroller");
			scrollerField.setAccessible(true);
			Field interpolatorField = ViewPager.class.getDeclaredField("sInterpolator");
			interpolatorField.setAccessible(true);

			scroller = new CustomDurationScroller(getContext(), (Interpolator) interpolatorField.get(null));
			scrollerField.set(this, scroller);
		} catch (Exception e) {
			e.printStackTrace();
		}
	}

	public void scrollOnce() {
		PagerAdapter adapter = getAdapter();
		int currentItem = getCurrentItem();
		int totalCount;
		if (adapter == null || (totalCount = adapter.getCount()) <= 1) {
			return;
		}

		int nextItem = (direction == LEFT) ? --currentItem : ++currentItem;
		if (nextItem < 0) {
			if (isCycle) {
				setCurrentItem(totalCount - 1, isBorderAnimation);
			}
		} else if (nextItem == totalCount) {
			if (isCycle) {
				setCurrentItem(0, isBorderAnimation);
			}
		} else {
			setCurrentItem(nextItem, true);
		}
	}

	@Override
	public boolean dispatchTouchEvent(MotionEvent ev) {
		int action = ev.getActionMasked();

		if (stopScrollWhenTouch) {
			if ((action == MotionEvent.ACTION_DOWN) && isAutoScroll) {
				isStopByTouch = true;
				stopAutoScroll();
			} else if (ev.getAction() == MotionEvent.ACTION_UP && isStopByTouch) {
				startAutoScroll();
			}
		}

		if (slideBorderMode == SLIDE_BORDER_MODE_TO_PARENT || slideBorderMode == SLIDE_BORDER_MODE_CYCLE) {
			touchX = ev.getX();
			if (ev.getAction() == MotionEvent.ACTION_DOWN) {
				downX = touchX;
			}
			int currentItem = getCurrentItem();
			PagerAdapter adapter = getAdapter();
			int pageCount = adapter == null ? 0 : adapter.getCount();

			if ((currentItem == 0 && downX <= touchX) || (currentItem == pageCount - 1 && downX >= touchX)) {
				if (slideBorderMode == SLIDE_BORDER_MODE_TO_PARENT) {
					getParent().requestDisallowInterceptTouchEvent(false);
				} else {
					if (pageCount > 1) {
						setCurrentItem(pageCount - currentItem - 1, isBorderAnimation);
					}
					getParent().requestDisallowInterceptTouchEvent(true);
				}
				return super.dispatchTouchEvent(ev);
			}
		}
		getParent().requestDisallowInterceptTouchEvent(true);

		return super.dispatchTouchEvent(ev);
	}

	private static class MyHandler extends Handler {

		private final WeakReference<StaxAutoScrollViewPager> autoScrollViewPager;

		public MyHandler(StaxAutoScrollViewPager staxAutoScrollViewPager) {
			this.autoScrollViewPager = new WeakReference<StaxAutoScrollViewPager>(staxAutoScrollViewPager);
		}

		@Override
		public void handleMessage(@NotNull Message msg) {
			super.handleMessage(msg);

			if (msg.what == SCROLL_WHAT) {
				StaxAutoScrollViewPager pager = this.autoScrollViewPager.get();
				if (pager != null) {
					pager.scroller.setScrollDurationFactor(pager.autoScrollFactor);
					pager.scrollOnce();
					pager.scroller.setScrollDurationFactor(pager.swipeScrollFactor);
					pager.sendScrollMessage(pager.interval + pager.scroller.getDuration());
				}
			}
		}
	}

	public void setStopScrollWhenTouch(boolean stopScrollWhenTouch) {
		this.stopScrollWhenTouch = stopScrollWhenTouch;
	}

	public void setCycle(boolean isCycle) {
		this.isCycle = isCycle;
	}

	public void setDirection(int direction) {
		this.direction = direction;
	}

	public void setInterval(long interval) {
		this.interval = interval;
	}

	public int getDirection() {
		return (direction == LEFT) ? LEFT : RIGHT;
	}

	private static class CustomDurationScroller extends Scroller {

		private double scrollFactor = 1;

		public CustomDurationScroller(Context context, Interpolator interpolator) {
			super(context, interpolator);
		}

		public void setScrollDurationFactor(double scrollFactor) {
			this.scrollFactor = scrollFactor;
		}


		@Override
		public void startScroll(int startX, int startY, int dx, int dy, int duration) {
			super.startScroll(startX, startY, dx, dy, (int) (duration * scrollFactor));
		}
	}

}

File: app/src/main/java/com/hover/stax/onboarding/slidingVariant/StaxAutoScrollViewPager.kt
package com.hover.stax.onboarding.slidingVariant

import android.content.Context
import android.os.Handler
import android.os.Message
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.animation.Interpolator
import android.widget.Scroller
import androidx.viewpager.widget.ViewPager
import java.lang.ref.WeakReference

class StaxAutoScrollViewPager : ViewPager {

    private var interval = DEFAULT_INTERVAL.toLong()
    private var direction = RIGHT
    private var isCycle = true
    private var stopScrollWhenTouch = true
    private val slideBorderMode = SLIDE_BORDER_MODE_NONE
    private val isBorderAnimation = true
    private var autoScrollFactor = 1.0
    private var swipeScrollFactor = 1.0
    private lateinit var handler: MyHandler
    private var isAutoScroll = false
    private var isStopByTouch = false
    private var touchX = 0f
    private var downX = 0f
    private var scroller: CustomDurationScroller? = null

    constructor(paramContext: Context?) : super(paramContext!!) {
        init()
    }

    constructor(paramContext: Context?, paramAttributeSet: AttributeSet?) : super(paramContext!!,
            paramAttributeSet) {
        init()
    }

    private fun init() {
        handler = MyHandler(this)
        setViewPagerScroller()
    }

    private fun startAutoScroll() {
        isAutoScroll = true
        sendScrollMessage((interval + scroller!!.duration / autoScrollFactor * swipeScrollFactor).toLong())
    }

    fun startAutoScroll(delayTimeInMills: Int) {
        isAutoScroll = true
        sendScrollMessage(delayTimeInMills.toLong())
    }

    private fun stopAutoScroll() {
        isAutoScroll = false
        handler.removeMessages(SCROLL_WHAT)
    }

    fun setSwipeScrollDurationFactor(scrollFactor: Double) {
        swipeScrollFactor = scrollFactor
    }

    fun setAutoScrollDurationFactor(scrollFactor: Double) {
        autoScrollFactor = scrollFactor
    }

    private fun sendScrollMessage(delayTimeInMills: Long) {
        handler.removeMessages(SCROLL_WHAT)
        handler.sendEmptyMessageDelayed(SCROLL_WHAT, delayTimeInMills)
    }

    private fun setViewPagerScroller() {
        try {
            val scrollerField = ViewPager::class.java.getDeclaredField("mScroller")
            scrollerField.isAccessible = true
            val interpolatorField = ViewPager::class.java.getDeclaredField("sInterpolator")
            interpolatorField.isAccessible = true
            scroller = CustomDurationScroller(context, interpolatorField[null] as Interpolator)
            scrollerField[this] = scroller
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    fun scrollOnce() {
        val adapter = adapter
        var currentItem = currentItem
        var totalCount = 0
        if (adapter == null || adapter.count.also { totalCount = it } <= 1) {
            return
        }
        val nextItem = if (direction == LEFT) --currentItem else ++currentItem
        if (nextItem < 0) {
            if (isCycle) {
                setCurrentItem(totalCount - 1, isBorderAnimation)
            }
        } else if (nextItem == totalCount) {
            if (isCycle) {
                setCurrentItem(0, isBorderAnimation)
            }
        } else {
            setCurrentItem(nextItem, true)
        }
    }

    private fun pauseScrolling(ev: MotionEvent) {
        if (ev.actionMasked == MotionEvent.ACTION_DOWN && isAutoScroll) {
            isStopByTouch = true
            stopAutoScroll()
        } else if (ev.action == MotionEvent.ACTION_UP && isStopByTouch) {
            startAutoScroll()
        }
    }

    override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
        if (stopScrollWhenTouch) pauseScrolling(ev)

        if (slideBorderMode == SLIDE_BORDER_MODE_TO_PARENT || slideBorderMode == SLIDE_BORDER_MODE_CYCLE) {
            touchX = ev.x
            if (ev.action == MotionEvent.ACTION_DOWN) {
                downX = touchX
            }
            val currentItem = currentItem
            val adapter = adapter
            val pageCount = adapter?.count ?: 0
            if (currentItem == 0 && downX <= touchX || currentItem == pageCount - 1 && downX >= touchX) {
                if (slideBorderMode == SLIDE_BORDER_MODE_TO_PARENT) {
                    parent.requestDisallowInterceptTouchEvent(false)
                } else {
                    if (pageCount > 1) {
                        setCurrentItem(pageCount - currentItem - 1, isBorderAnimation)
                    }
                    parent.requestDisallowInterceptTouchEvent(true)
                }
                return super.dispatchTouchEvent(ev)
            }
        }
        parent.requestDisallowInterceptTouchEvent(true)
        return super.dispatchTouchEvent(ev)
    }

    private class MyHandler(staxAutoScrollViewPager: StaxAutoScrollViewPager) : Handler() {
        private val autoScrollViewPager: WeakReference<StaxAutoScrollViewPager> = WeakReference(staxAutoScrollViewPager)
        override fun handleMessage(msg: Message) {
            super.handleMessage(msg)
            if (msg.what == SCROLL_WHAT) {
                val pager = autoScrollViewPager.get()
                if (pager != null) {
                    pager.scroller!!.setScrollDurationFactor(pager.autoScrollFactor)
                    pager.scrollOnce()
                    pager.scroller!!.setScrollDurationFactor(pager.swipeScrollFactor)
                    pager.sendScrollMessage(pager.interval + pager.scroller!!.duration)
                }
            }
        }

    }

    fun setStopScrollWhenTouch(stopScrollWhenTouch: Boolean) {
        this.stopScrollWhenTouch = stopScrollWhenTouch
    }

    fun setCycle(isCycle: Boolean) {
        this.isCycle = isCycle
    }

    fun setDirection(direction: Int) {
        this.direction = direction
    }

    fun setInterval(interval: Long) {
        this.interval = interval
    }

    fun getDirection(): Int {
        return if (direction == LEFT) LEFT else RIGHT
    }

    private class CustomDurationScroller(context: Context?, interpolator: Interpolator?) :
            Scroller(context, interpolator) {
        private var scrollFactor = 1.0
        fun setScrollDurationFactor(scrollFactor: Double) {
            this.scrollFactor = scrollFactor
        }

        override fun startScroll(startX: Int, startY: Int, dx: Int, dy: Int, duration: Int) {
            super.startScroll(startX, startY, dx, dy, (duration * scrollFactor).toInt())
        }
    }

    companion object {
        const val DEFAULT_INTERVAL = 1500
        const val LEFT = 0
        const val RIGHT = 1
        const val SLIDE_BORDER_MODE_NONE = 0
        const val SLIDE_BORDER_MODE_CYCLE = 1
        const val SLIDE_BORDER_MODE_TO_PARENT = 2
        const val SCROLL_WHAT = 0
    }
}